lib/web_ui/dev/webdriver_browser.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:image/image.dart';
+import 'package:webdriver/io.dart' show createDriver, WebDriver;
+
+import 'browser.dart';
+
+abstract class WebDriverBrowserEnvironment extends BrowserEnvironment {
+  late final Process _driverProcess;
+
+  Future<Process> spawnDriverProcess();
+  Uri get driverUri;
+
+  @override
+  Future<void> prepare() async {
+    _driverProcess = await spawnDriverProcess();
+
+    _driverProcess.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String error) {
+      print('[Webdriver][Error] $error');
+    });
+
+    _driverProcess.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String log) {
+      print('[Webdriver] $log');
+    });
+  }
+
+  @override
+  Future<void> cleanup() async {
+    _driverProcess.kill();
+  }
+
+  @override
+  Future<Browser> launchBrowserInstance(Uri url, {bool debug = false}) async {
+    for (;;) {
+      try {
+        final WebDriver driver = await createDriver(
+          uri: driverUri, desired: <String, dynamic>{'browserName': packageTestRuntime.identifier});
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
+        // Sometimes we may try to connect before the web driver port is ready.
+        // So we should here. Note that if there was some issue with the
+        // webdriver process, we may loop infinitely here, so we're relying on
+        // the test timeout to kill us if it takes too long to connect.
+        print('Failed to connect to webdriver process. Retrying in 100 ms');
+        await Future<void>.delayed(const Duration(milliseconds: 100));
+      } catch (exception) {
+        print('Exception while creating webdriver: $exception');
```

Comment:
When rethrowing you don't need to also print, because the `rethrow` will retain all the information about the original error, including its stack trace.